### PR TITLE
feat: Navigation nodes with a page attached use the preview url in the edit or preview mode.

### PR DESCRIPTION
**Pull request by Aiky30**
The navigation will output the preview url for any navigation nodes that have a page attached in the edit or preview mode.

### CODE REVIEW

File: djangocms_navigation/utils.py [hunk 1]
```diff
@@ -69,3 +73,33 @@ def get_versionable_for_content(content):
 
 def purge_menu_cache(site_id=None, language=None):
     menu_pool.clear(site_id=site_id, language=language)
+
+
+def is_preview_or_edit_mode(request):
+    """
+    Detemrine if the view is in the preview or edit mode.
```
Comment (Aiky30) [inline djangocms_navigation/utils.py:80]:
```suggestion
    Determine if the view is in the preview or edit mode.
```

File: CHANGELOG.rst [hunk 2]
```diff
@@ -4,6 +4,7 @@ Changelog
 
 Unreleased
 ==========
+* feat: The navigation will output the preview url for any navigation nodes that have a page attached in the edit or preview mode.
```
Comment (Aiky30) [inline CHANGELOG.rst:7]:
```suggestion
* feat: Navigation nodes with a page attached use the preview url in the edit or preview mode. 
```

File: djangocms_navigation/utils.py [hunk 3]
```diff
@@ -69,3 +73,33 @@ def get_versionable_for_content(content):
...
+    """
+    toolbar = getattr(request, "toolbar", None)
+    if toolbar and (toolbar.edit_mode_active or toolbar.preview_mode_active):
+        return True
+    return False
+
+
+def get_latest_page_content_for_page_grouper(obj, language):
+    """
+    Determine if the view is in the preview or edit mode.
```
Comment (Aiky30) [inline djangocms_navigation/utils.py:94]:
```suggestion
    Get the latest draft or published page content from a page grouper object
```

**Review by michaeljcollinsuk — APPROVED** [review]
I'm still unfamiliar with the `djangocms-navigation` package, but the changes themselves look good to me

**Review by Bernardvdv — APPROVED** [review]

**Comment by michaeljcollinsuk** [comment]
> > There is quite a lot of repetition between the tests, I wondered if you would consider using subtests or if there is a reason you prefer not to?
> 
> @michaeljcollinsuk I prefer to separate concerns so that each test controls what it's creating rather than one implementation touching many tests.
> 
> We sometimes use the setup method to pre create the test data, it then gets a bit ugly if you want to add more tests because you could then be creating content that you don't need in the same TestCase class, the same for if anyone then modifies the values in setup / shared logic which could impact your test.
> 
> I've not used subtests, I have seen the odd one in the core of the cms. I think this could then suffer from the same issue as shared data in setup where someone wants to use it again and then changes it slightly, it then impacts other tests. I tend to look at tests from a maintenance perspective, the less "sharing" in my opinion makes maintenance easier, especially if someone changes things to the point where other tests are then broken but still passing (core of the cms also suffers badly from this).
> 
> The factories should probably be updated to cover creating a menu with a page node which would clean it up a fair bit. Open to suggestions.

Fair points, I agree with what you say about the dangers of doing too much in the setup method. I think using that works best when the TestCase classes are small and focussed. And also agree that when it comes to tests, I would rather have them easier to follow and maintain even if this leads to some repetition.

As for using subtests, I do like them but it does depend on the circumstances. But as an example of how I might write these tests as a single test using subtest:
```
    def test_page_menu_links(self):
        preview_pagecontent = factories.PageContentWithVersionFactory(
            language=self.language,
            version__created_by=self.get_superuser(),
            version__state=PUBLISHED,
        )
        menu = factories.MenuFactory()
        menu_content = factories.MenuContentWithVersionFactory(
            menu=menu,
            version__state=PUBLISHED,
            language=self.language
        )
        menu_item_pagecontent = factories.PageContentWithVersionFactory(
            language=self.language,
            version__created_by=self.get_superuser(),
            version__state=PUBLISHED,
        )
        factories.ChildMenuItemFactory(parent=menu_content.root, content=menu_item_pagecontent.page)
        preview_url = get_object_preview_url(menu_item_pagecontent, language=self.language)
        live_url = menu_item_pagecontent.get_absolute_url(language=self.language)

        # testcases
        test_cases = [
            {
                "name": "edit_endpoint",
                "endpoint": get_object_edit_url(preview_pagecontent, language=self.language),
                "expected": preview_url,
                "not_expected": live_url,
            },
            {
                "name": "preview_endpoint",
                "endpoint": get_object_preview_url(preview_pagecontent, language=self.language),
                "expected": preview_url,
                "not_expected": live_url,
            },
            {
                "name": "live_endpoint",
                "endpoint": preview_pagecontent.get_absolute_url(language=self.language),
                "expected": live_url,
                "not_expected": preview_url,
            },
        ]

        for test_case in test_cases:
            with self.subTest(msg=test_case["name"], test_case=test_case):

                with self.login_user_context(self.get_superuser()):
                    response = self.client.get(test_case["endpoint"])

                self.assertEqual(response.status_code, 200)

                soup = BeautifulSoup(str(response.content), features="lxml")
                navigation_tree = soup.find("ul", class_="nav")
                first_child_link = navigation_tree.find('a', href=True)['href']

                self.assertIn(test_case["expected"], first_child_link)
                self.assertNotIn(test_case["not_expected"], first_child_link)

```

**Comment by Aiky30** [comment]
> > > There is quite a lot of repetition between the tests, I wondered if you would consider using subtests or if there is a reason you prefer not to?
> > 
> > 
> > @michaeljcollinsuk I prefer to separate concerns so that each test controls what it's creating rather than one implementation touching many tests.
> > We sometimes use the setup method to pre create the test data, it then gets a bit ugly if you want to add more tests because you could then be creating content that you don't need in the same TestCase class, the same for if anyone then modifies the values in setup / shared logic which could impact your test.
> > I've not used subtests, I have seen the odd one in the core of the cms. I think this could then suffer from the same issue as shared data in setup where someone wants to use it again and then changes it slightly, it then impacts other tests. I tend to look at tests from a maintenance perspective, the less "sharing" in my opinion makes maintenance easier, especially if someone changes things to the point where other tests are then broken but still passing (core of the cms also suffers badly from this).
> > The factories should probably be updated to cover creating a menu with a page node which would clean it up a fair bit. Open to suggestions.
> 
> Fair points, I agree with what you say about the dangers of doing too much in the setup method. I think using that works best when the TestCase classes are small and focussed. And also agree that when it comes to tests, I would rather have them easier to follow and maintain even if this leads to some repetition.
> 
> As for using subtests, I do like them but it does depend on the circumstances. But as an example of how I might write these tests as a single test using subtest:
> 
> ```
>     def test_page_menu_links(self):
>         preview_pagecontent = factories.PageContentWithVersionFactory(
>             language=self.language,
>             version__created_by=self.get_superuser(),
>             version__state=PUBLISHED,
>         )
>         menu = factories.MenuFactory()
>         menu_content = factories.MenuContentWithVersionFactory(
>             menu=menu,
>             version__state=PUBLISHED,
>             language=self.language
>         )
>         menu_item_pagecontent = factories.PageContentWithVersionFactory(
>             language=self.language,
>             version__created_by=self.get_superuser(),
>             version__state=PUBLISHED,
>         )
>         factories.ChildMenuItemFactory(parent=menu_content.root, content=menu_item_pagecontent.page)
>         preview_url = get_object_preview_url(menu_item_pagecontent, language=self.language)
>         live_url = menu_item_pagecontent.get_absolute_url(language=self.language)
> 
>         # testcases
>         test_cases = [
>             {
>                 "name": "edit_endpoint",
>                 "endpoint": get_object_edit_url(preview_pagecontent, language=self.language),
>                 "expected": preview_url,
>                 "not_expected": live_url,
>             },
>             {
>                 "name": "preview_endpoint",
>                 "endpoint": get_object_preview_url(preview_pagecontent, language=self.language),
>                 "expected": preview_url,
>                 "not_expected": live_url,
>             },
>             {
>                 "name": "live_endpoint",
>                 "endpoint": preview_pagecontent.get_absolute_url(language=self.language),
>                 "expected": live_url,
>                 "not_expected": preview_url,
>             },
>         ]
> 
>         for test_case in test_cases:
>             with self.subTest(msg=test_case["name"], test_case=test_case):
> 
>                 with self.login_user_context(self.get_superuser()):
>                     response = self.client.get(test_case["endpoint"])
> 
>                 self.assertEqual(response.status_code, 200)
> 
>                 soup = BeautifulSoup(str(response.content), features="lxml")
>                 navigation_tree = soup.find("ul", class_="nav")
>                 first_child_link = navigation_tree.find('a', href=True)['href']
> 
>                 self.assertIn(test_case["expected"], first_child_link)
>                 self.assertNotIn(test_case["not_expected"], first_child_link)
> ```

Thanks for putting the time in to create that. I'm not keen on how you could make a minor change for one scenario and it could break others with out being obvious. The core cms has some tests like this and they are really hard to make changes to because you don't have enough information as to what it's testing and why, it's like a black box. 

You were right to point this out because it was very messy. I've cleaned it up to a standard that I should have created first where it uses some setup, reusable functions and hopefully only has what's different in each test. Happy for any feedback / requests to change.

File: tests/test_cms_menus.py [hunk 4]
```diff
@@ -391,6 +392,158 @@ def test_published_menu_in_preview_mode(self):
...
+        self.assertNotIn(not_expected_live_endpoint, actual_link)
+
+    def test_archived_page_in_menu_links_in_preview_mode(self):
+        """
+        The preview should show no link for any menu items attached to an archived page
+        """
+        menu_item_pagecontent = factories.PageContentWithVersionFactory(
+            language=self.language,
+            version__created_by=self.get_superuser(),
+            version__state=UNPUBLISHED,
```
Comment (michaeljcollinsuk) [inline tests/test_cms_menus.py:529]: Should this be `ARCHIVED`?

Comment (Aiky30) [reply tests/test_cms_menus.py:529]: As this is a factory it's already a method of sorts. We are only passing in what we want to control.

Comment (Aiky30) [reply tests/test_cms_menus.py:529]: It should have been ARCHIVED, well spotted. GitHuh has a really cool suggestions tool for small suggestions like this, I'll even use it myself for simplicity sometimes.

```suggestion
            version__state=ARCHIVED,
```

File: tests/test_utils.py [hunk 5]
```diff
@@ -84,3 +96,84 @@ def test_non_supported_content_types(self):
         unexpected_content_types = ContentType.objects.get_for_models(User)
         for model in unexpected_content_types:
             self.assertFalse(is_model_supported(MenuContent, model))
+
+
+class CMSMenuTestCase(CMSTestCase):
```
Comment (Aiky30) [inline tests/test_utils.py:101]:
```suggestion
class IsPreviewEditModeTestCase(CMSTestCase):
```

**Pull request merged**